react-front-end/tsrc/modules/BrowserStorageModule.ts

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to The Apereo Foundation under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * The Apereo Foundation licenses this file to you under the Apache License,
+ * Version 2.0, (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as E from "fp-ts/Either";
+import { pipe, flow } from "fp-ts/function";
+import * as J from "fp-ts/Json";
+import * as O from "fp-ts/Option";
+import { AppConfig } from "../AppConfig";
+
+const localStorage = window.localStorage;
+
+const buildStorageKey = (key: string) => `${AppConfig.baseUrl}_${key}`;
+
+/**
+ * Retrieve data from browser local storage by a key. Return undefined if
+ * no data is found.
+ *
+ * @param key Key of the data to be read.
+ * @param validator A function to perform type checking against the parsed value.
+ */
+export const readDataFromLocalStorage = <T>(
+  key: string,
+  validator: (value: unknown) => value is T
+): T | undefined =>
+  pipe(
+    localStorage.getItem(buildStorageKey(key)),
+    O.fromNullable,
+    O.fold(
+      () => undefined,
+      flow(
+        J.parse,
+        E.mapLeft((error) => `Failed to parse data due to ${error}`),
+        E.filterOrElse(
+          validator,
+          () => "Data format mismatch with data read from storage"
+        ),
+        E.mapLeft(console.error),
+        E.getOrElseW(() => undefined)
```

Comment:
With the above suggestion, perhaps this line should just be:

```suggestion
        O.fromEither)
```

Mind you, in doing that, you'd then want to use `O.chain` rather than `O.map`.

But at least with these two comments we only go back to the wild west of `undefined` and `null` when we exit the functional area.